include/zephyr/drivers/usb/udc_buf.h

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief Buffers for USB device support
+ */
+
+#ifndef ZEPHYR_INCLUDE_UDC_BUF_H
+#define ZEPHYR_INCLUDE_UDC_BUF_H
+
+#include <zephyr/kernel.h>
+#include <zephyr/net/buf.h>
+
+#if defined(CONFIG_DCACHE) && !defined(CONFIG_UDC_BUF_FORCE_NOCACHE)
+/*
+ * Here we try to get DMA-safe buffers, but we lack a consistent source of
+ * information about data cache properties, such as line cache size, and a
+ * consistent source of information about what part of memory is DMA'able.
+ * For now, we simply assume that all available memory is DMA'able and use
+ * Kconfig option DCACHE_LINE_SIZE for alignment and granularity.
+ */
+#define Z_UDC_BUF_ALIGN		CONFIG_DCACHE_LINE_SIZE
+#define Z_UDC_BUF_GRANULARITY	CONFIG_DCACHE_LINE_SIZE
+#else
+/*
+ * Default alignment and granularity to pointer size if the platform does not
+ * have a data cache or buffers are placed in nocache memory region.
+ */
+#define Z_UDC_BUF_ALIGN		sizeof(void *)
+#define Z_UDC_BUF_GRANULARITY	sizeof(void *)
+#endif
+
+/**
+ * @brief Buffer API for USB device support
+ * @defgroup udc_buf Experimental buffer API for USB device support
+ * @ingroup usb
+ * @{
+ */
+
+/** Buffer alignment required by the UDC driver */
+#define UDC_BUF_ALIGN		Z_UDC_BUF_ALIGN
+
+/** Buffer granularity required by the UDC driver */
+#define UDC_BUF_GRANULARITY	Z_UDC_BUF_GRANULARITY
+
+/**
+ * @brief Helper macro for specifying buffer alignment.
+ *
+ * This macro should be used if the application defines its own buffers to be
+ * used for USB transfers.
+ *
+ * Example use to make buffer alignment UDC compliant:
+ *
+ * @code{.c}
+ *     static uint8_t udc_align_attr report[KB_REPORT_COUNT];
```

Comment:
This example would need the `KB_REPORT_COUNT` to be increased to be multiple of `UDC_BUF_GRANULARITY`, e.g. `ROUND_UP(KB_REPORT_COUNT, UDC_BUF_GRANULARITY)`.

Reply:
Yeah, I have played with doxygen, still trying to figure out how to make it better. Maybe some macro to define static and compliant buffer.